### DISCUSSION TITLE
Is there a way to improve the compilation speed of `main` process code (development)?

### DISCUSSION THREAD

**Discussion by 1111mp** [alex8088/electron-vite#404] [Q&A]
This is my current configuration `electron.vite.config.ts`:
```ts
import { resolve } from "path";
import { defineConfig, externalizeDepsPlugin } from "electron-vite";
import react from "@vitejs/plugin-react";
import del from "rollup-plugin-delete";
import swc from "@rollup/plugin-swc";

const isTest = process.env.TEST === "true";

export default defineConfig({
  main: {
    resolve: {
      alias: {
        "@src": resolve("src")
      }
    },
    build: {
      rollupOptions: {
        input: isTest ? "src/main/main.test.ts" : "src/main/main.ts",
        output: {
          entryFileNames: "main.mjs",
          inlineDynamicImports: true,
          format: "es"
        },
        plugins: [swc()]
      }
    },
    plugins: [del({ targets: ["release/app/dist", "release/build"] })]
  },
  preload: {
    build: {
      rollupOptions: {
        input: isTest ? "src/preload/preload.test.ts" : "src/preload/index.ts",
        output: {
          entryFileNames: "preload.js",
          inlineDynamicImports: true,
          format: "cjs"
        }
      }
    },
    plugins: [externalizeDepsPlugin()]
  },
  renderer: {
    resolve: {
      alias: {
        "@src": resolve("src"),
        "@renderer": resolve("src/renderer/src")
      }
    },
    plugins: [react()]
  }
});
```

I tried using `swc` to speed up the build but it didn't work. I also tried [`swcPlugin`](https://electron-vite.org/guide/typescript-decorator#typescript-decorator).

Currently it takes about two seconds to complete:
```shell
> electron-vite dev --watch

vite v5.0.12 building SSR bundle for development...

watching for file changes...

build started...
✓ 990 modules transformed.
out/main/main.mjs  2,047.89 kB
built in 2141ms.

build the electron main process successfully
```

Maybe I'm using it the wrong way, please feel free to correct me.

I tried using [`tsup`](https://tsup.egoist.dev/) (it is based on `esbuild`), which can greatly improve the build speed. It only takes about 100 milliseconds in the end. But this would deviate from `vite`. So I want to know if there is a better way to do the same job as using `esbuild`. Thanks in advance. 🌹 

`tsup.config.ts`:
```ts
import { defineConfig } from "tsup";

export default defineConfig({
  clean: true,
  entry: ["src/main/main.ts"],
  target: "esnext",
  format: ["esm"],
  bundle: true,
  external: ["electron"],
  outDir: "out/main",
  inject: ["cjs-shim.ts"],
  tsconfig: "tsconfig.node.json"
});
```

**Comment by alex8088** [comment]
https://electron-vite.org/guide/dev#dependencies-vs-devdependencies
enable the externalizedep plugin

**Comment by 1111mp** [comment]
I have read the documentation and carefully checked the source code. This approach seems to only improve the build speed by reducing the external dependencies that need to be compiled through `external`, but it is limited when using `prod build` (only `electron` & `node` modules can be excluded).

I think I've found a solution that works for me. I `fork` a copy of the code locally and successfully rewrote it to use `esbuild` to compile the code (only relying on `vite`’s `devServer`). Maybe I should make the relevant code public in the future so that people with similar needs can view it.

Thank you very much for your reply though.  Have a nice day~ 🌹

**Comment by alex8088** [comment]
no, you should enable externalizedep plugin that will externalize all dependences that you install in deps, like no bundle.

**Comment by 1111mp** [comment]
Yes I know that. 

When I enable`externalizeDepsPlugin`:
<img width="505" alt="Screenshot 2024-02-08 at 22 35 53" src="https://github.com/alex8088/electron-vite/assets/31227919/38465052-95bc-45db-a1b0-8bd2aebd9c88">

it does get faster. But this doesn't work for compiling the production package,because as the project progresses, some external dependencies must be packaged in.

When I use `esbuild`:
<img width="544" alt="Screenshot 2024-02-08 at 22 40 28" src="https://github.com/alex8088/electron-vite/assets/31227919/8e439bd3-2a9c-49bd-bdcb-13dd0603bf89">

More importantly, the performance of compiling the production package is also the same.
